services/gossip/adapter/tcp/client_test.go

```diff
@@ -0,0 +1,127 @@
+// Copyright 2019 the orbs-network-go authors
+// This file is part of the orbs-network-go library in the Orbs project.
+//
+// This source code is licensed under the MIT license found in the LICENSE file in the root directory of this source tree.
+// The above notice should be included in all copies or substantial portions of the software.
+
+package tcp
+
+import (
+	"context"
+	"github.com/orbs-network/orbs-network-go/config"
+	"github.com/orbs-network/orbs-network-go/instrumentation/metric"
+	"github.com/orbs-network/orbs-network-go/test"
+	"github.com/orbs-network/scribe/log"
+	"github.com/stretchr/testify/require"
+	"net"
+	"testing"
+	"time"
+)
+
+func TestClientConnection_EnablesQueueWhenConnectedToServer_AndDisablesQueueOnDisconnect(t *testing.T) {
+	test.WithContext(func(ctx context.Context) {
+		server := newServerStub(t)
+		defer server.Close()
+
+		client := server.createClient(ctx, t, 20*time.Hour) // so that we don't send keep alives
+
+		waitForQueueEnabled(t, client)
+
+		client.disconnect()
+
+		waitForQueueDisabled(t, client)
+
+		require.Zero(t, server.readSomeBytes(), "client shouldn't have sent anything")
+	})
+}
+
+func TestClientConnection_ReconnectsWhenServerDisconnects_AndSendKeepAlive(t *testing.T) {
+	test.WithContext(func(ctx context.Context) {
+		server := newServerStub(t)
+		defer server.Close()
+
+		client := server.createClient(ctx, t, 10*time.Millisecond)
+		waitForQueueEnabled(t, client)
+
+		server.forceDisconnect(t)
```

Comment:
I wanted to do that but it's too difficult; my test harness disconnects by closing the connection, but I think that it doesn't prevent reconnecting - the reconnection attempt just blocks until we call `Accept` again.

Do you think it's worth investing time in this? we're checking that the queue is disabled in `TestClientConnection_EnablesQueueWhenConnectedToServer_AndDisablesQueueOnDisconnect` and the queue is disabled on exiting the connection handler in all code paths

Reply:
1. The concern is that the disconnect would take place after the test already passes. since both `readSomeBytes` and `forceDisconnect` work in synchronous calls against the connection object this is possibly not a problem (they are assumed to enforce strict sequential access to the resource. but suppose the connection's buffer allows you to read previously unread  bytes after the connection has already been closed - in this case, it's possible that the two KA were sent before `forceDisconnect` ever happened. Since the queues are never disabled after the server closes the connection, waitForQueueEnabled has no meaning here, and the test would pass even though the KA were sent before the disconnect

2. We're missing a test (and I believe it would fail if we had it) to check that the queues become disabled after the server kills the connection. `TestClientConnection_EnablesQueueWhenConnectedToServer_AndDisablesQueueOnDisconnect` only checks the case where the disconnect is caused by client context shutdown.

3. If we had a test (and support) for the case in &2 I think you would be safe in adding `		waitForQueueDisabled(t, client)` after `forceDisconnect` making this test more bulletproof